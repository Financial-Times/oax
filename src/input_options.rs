use fnv::FnvHashSet;

#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub enum PackageManager {
    Bower,
    Npm,
}

impl PackageManager {
    pub fn dir(&self) -> &'static str {
        match self {
            Self::Bower => "bower_components",
            Self::Npm => "node_modules",
        }
    }

    pub fn file(&self) -> &'static str {
        match self {
            Self::Bower => ".bower.json",
            Self::Npm => "package.json",
        }
    }
}

impl Default for PackageManager {
    fn default() -> PackageManager {
        PackageManager::Npm
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct InputOptions {
    pub package_manager: PackageManager,
    pub external: FnvHashSet<String>,
}
