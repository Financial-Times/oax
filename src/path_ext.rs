use matches::matches;
use std::mem;
use std::path::{self, Path, PathBuf};

pub trait PathBufExt {
    fn append_resolving<P: AsRef<Path> + ?Sized>(&mut self, more: &P);
    fn prepend_resolving<P: AsRef<Path> + ?Sized>(&mut self, base: &P);
    fn empty(&mut self);
    fn replace_with<P: AsRef<Path> + ?Sized>(&mut self, that: &P);
    fn as_mut_vec(&mut self) -> &mut Vec<u8>;
}

impl PathBufExt for PathBuf {
    fn append_resolving<P: AsRef<Path> + ?Sized>(&mut self, more: &P) {
        for c in more.as_ref().components() {
            match c {
                path::Component::Prefix(prefix) => {
                    *self = PathBuf::from(prefix.as_os_str().to_owned());
                }
                path::Component::RootDir => {
                    self.push(path::MAIN_SEPARATOR.to_string());
                }
                path::Component::CurDir => {}
                path::Component::ParentDir => {
                    self.pop();
                }
                path::Component::Normal(part) => {
                    self.push(part);
                }
            }
        }
    }
    fn prepend_resolving<P: AsRef<Path> + ?Sized>(&mut self, base: &P) {
        let mut tmp = base.as_ref().to_owned();
        mem::swap(self, &mut tmp);
        self.append_resolving(tmp.as_path());
    }
    fn empty(&mut self) {
        self.as_mut_vec().clear();
    }
    fn replace_with<P: AsRef<Path> + ?Sized>(&mut self, that: &P) {
        self.empty();
        self.push(that);
    }

    fn as_mut_vec(&mut self) -> &mut Vec<u8> {
        unsafe { &mut *(self as *mut PathBuf as *mut Vec<u8>) }
    }
}

pub trait PathExt {
    fn is_explicitly_relative(&self) -> bool;
    fn relative_from<P: AsRef<Path> + ?Sized>(&self, base: &P) -> Option<PathBuf>;
}

impl PathExt for Path {
    #[inline]
    fn is_explicitly_relative(&self) -> bool {
        matches!(
            self.components().next(),
            Some(path::Component::CurDir) | Some(path::Component::ParentDir)
        )
    }
    fn relative_from<P: AsRef<Path> + ?Sized>(&self, base: &P) -> Option<PathBuf> {
        let base = base.as_ref();
        if self.is_absolute() != base.is_absolute() {
            if self.is_absolute() {
                Some(PathBuf::from(self))
            } else {
                None
            }
        } else {
            let mut ita = self.components();
            let mut itb = base.components();
            let mut comps: Vec<path::Component> = vec![];
            loop {
                match (ita.next(), itb.next()) {
                    (None, None) => break,
                    (Some(a), None) => {
                        comps.push(a);
                        comps.extend(ita.by_ref());
                        break;
                    }
                    (None, _) => comps.push(path::Component::ParentDir),
                    (Some(a), Some(b)) if comps.is_empty() && a == b => (),
                    (Some(a), Some(b)) if b == path::Component::CurDir => comps.push(a),
                    (Some(_), Some(b)) if b == path::Component::ParentDir => return None,
                    (Some(a), Some(_)) => {
                        comps.push(path::Component::ParentDir);
                        for _ in itb {
                            comps.push(path::Component::ParentDir);
                        }
                        comps.push(a);
                        comps.extend(ita.by_ref());
                        break;
                    }
                }
            }
            Some(comps.iter().map(|c| c.as_os_str()).collect())
        }
    }
}
