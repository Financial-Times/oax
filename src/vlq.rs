use std::str;

const B64: &[u8] = b"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

pub struct Vlq {
    buf: [u8; 13],
}

impl Vlq {
    pub fn new() -> Self {
        Self { buf: [0u8; 13] }
    }

    pub fn enc(&mut self, n: isize) -> &str {
        let sign = n < 0;
        let n = if sign { -n } else { n } as usize;
        let mut y = (n & 0xf) << 1 | sign as usize;
        let mut r = n >> 4;
        let mut l = 0;
        while r > 0 {
            y |= 0x20;
            self.buf[l] = B64[y];
            y = r & 0x1f;
            r >>= 5;
            l += 1;
        }
        self.buf[l] = B64[y];
        str::from_utf8(&self.buf[0..=l]).unwrap()
    }
}
