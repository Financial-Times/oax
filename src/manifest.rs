use crate::path_ext::*;
use crate::CliError;
use fnv::FnvHashMap;
use matches::matches;
use serde::de::{SeqAccess, Visitor};
use serde::{de, Deserialize, Deserializer};
use std::cell::RefCell;
use std::marker::PhantomData;
use std::path::{Path, PathBuf};
use std::rc::Rc;
use std::{fmt, fs, io, mem};

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum BrowserSubstitution<T> {
    Ignore,
    Replace(T),
}

#[derive(Debug, Default, Deserialize, PartialEq, Eq, Clone)]
#[serde(transparent)]
pub struct BrowserSubstitutionMap(pub FnvHashMap<PathBuf, BrowserSubstitution<PathBuf>>);

#[derive(Debug, Clone, Default)]
pub struct PackageCache {
    pub pkgs: RefCell<FnvHashMap<PathBuf, Option<Rc<PackageInfo>>>>,
}

impl PackageCache {
    pub fn nearest_package_info(
        &self,
        mut dir: PathBuf,
        package_manager: crate::input_options::PackageManager,
    ) -> Result<Option<Rc<PackageInfo>>, CliError> {
        loop {
            if !matches!(dir.file_name(), Some(s) if s == package_manager.dir()) {
                if let Some(info) = self.package_info(&mut dir, package_manager)? {
                    return Ok(Some(info));
                }
            }
            if !dir.pop() {
                return Ok(None);
            }
        }
    }

    pub fn package_info(
        &self,
        dir: &mut PathBuf,
        package_manager: crate::input_options::PackageManager,
    ) -> Result<Option<Rc<PackageInfo>>, CliError> {
        let mut pkgs = self.pkgs.borrow_mut();
        let manifest_file_names = package_manager.files();
        Ok(pkgs
            .entry(dir.clone())
            .or_insert_with(|| {
                for manifest_file_name in manifest_file_names {
                    dir.push(manifest_file_name);
                    if let Ok(file) = fs::File::open(&dir) {
                        let buf_reader = io::BufReader::new(file);
                        if let Ok(mut info) = serde_json::from_reader::<_, PackageInfo>(buf_reader)
                        {
                            dir.pop();
                            info.set_base(&dir);
                            return Some(Rc::new(info));
                        } else {
                            dir.pop();
                        }
                    } else {
                        dir.pop();
                    }
                }
                None
            })
            .as_ref()
            .cloned())
    }
}

#[derive(Debug, Default, PartialEq, Eq, Clone)]
pub struct PackageInfo {
    pub main: PathBuf,
    pub browser_substitutions: BrowserSubstitutionMap,
}

impl PackageInfo {
    fn set_base(&mut self, base: &Path) {
        self.main.prepend_resolving(base);
        let substs = mem::replace(&mut self.browser_substitutions, Default::default());
        self.browser_substitutions
            .0
            .extend(substs.0.into_iter().map(|(mut from, mut to)| {
                if from.is_explicitly_relative() {
                    from.prepend_resolving(base);
                }
                match to {
                    BrowserSubstitution::Ignore => {}
                    BrowserSubstitution::Replace(ref mut path) => {
                        if path.is_explicitly_relative() {
                            path.prepend_resolving(base);
                        }
                    }
                }
                (from, to)
            }));
    }
}
impl<'de> Deserialize<'de> for PackageInfo {
    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
        #[derive(Debug, Default, PartialEq, Eq, Deserialize)]
        #[serde(default)]
        struct RawPackageInfo {
            #[serde(deserialize_with = "from_main")]
            main: Option<PathBuf>,
            browser: BrowserField,
        }
        let info = RawPackageInfo::deserialize(deserializer)?;
        let main = info.main.unwrap_or(PathBuf::from("./index"));
        let browser_substitutions = info.browser.to_map(&main);
        Ok(PackageInfo {
            main,
            browser_substitutions,
        })
    }
}

#[macro_export]
macro_rules! map {
    {} => {
        Default::default()
    };
    { $($key:expr => $value:expr,)+ } => {
        {
            let mut map = FnvHashMap::default();
            $(map.insert($key, $value);)+
            map
        }
    };
    { $($key:expr => $value:expr),+ } => {
        map!{$($key => $value,)+}
    };
}

macro_rules! visit_unconditionally {
    // bool i64 i128 u64 u128 f64 str bytes none some unit newtype_struct seq map enum
    ($l:lifetime $as:expr) => {};
    ($l:lifetime $as:expr, ) => {};
    ($l:lifetime $as:expr, bool $($x:tt)*) => {
        fn visit_bool<E: de::Error>(self, _: bool) -> Result<Self::Value, E> { Ok($as) }
        visit_unconditionally!($l $as, $($x)*);
    };
    ($l:lifetime $as:expr, i64 $($x:tt)*) => {
        fn visit_i64<E: de::Error>(self, _: i64) -> Result<Self::Value, E> { Ok($as) }
        visit_unconditionally!($l $as, $($x)*);
    };
    ($l:lifetime $as:expr, i128 $($x:tt)*) => {
        fn visit_i128<E: de::Error>(self, _: i128) -> Result<Self::Value, E> { Ok($as) }
        visit_unconditionally!($l $as, $($x)*);
    };
    ($l:lifetime $as:expr, u64 $($x:tt)*) => {
        fn visit_u64<E: de::Error>(self, _: u64) -> Result<Self::Value, E> { Ok($as) }
        visit_unconditionally!($l $as, $($x)*);
    };
    ($l:lifetime $as:expr, u128 $($x:tt)*) => {
        fn visit_u128<E: de::Error>(self, _: u128) -> Result<Self::Value, E> { Ok($as) }
        visit_unconditionally!($l $as, $($x)*);
    };
    ($l:lifetime $as:expr, f64 $($x:tt)*) => {
        fn visit_f64<E: de::Error>(self, _: f64) -> Result<Self::Value, E> { Ok($as) }
        visit_unconditionally!($l $as, $($x)*);
    };
    ($l:lifetime $as:expr, str $($x:tt)*) => {
        fn visit_str<E: de::Error>(self, v: &str) -> Result<Self::Value, E> { Ok($as) }
        visit_unconditionally!($l $as, $($x)*);
    };
    ($l:lifetime $as:expr, bytes $($x:tt)*) => {
        fn visit_bytes<E: de::Error>(self, _: &[u8]) -> Result<Self::Value, E> { Ok($as) }
        visit_unconditionally!($l $as, $($x)*);
    };
    ($l:lifetime $as:expr, none $($x:tt)*) => {
        fn visit_none<E: de::Error>(self) -> Result<Self::Value, E> { Ok($as) }
        visit_unconditionally!($l $as, $($x)*);
    };
    ($l:lifetime $as:expr, some $($x:tt)*) => {
        fn visit_some<D: Deserializer<$l>>(self, _: D) -> Result<Self::Value, D::Error> { Ok($as) }
        visit_unconditionally!($l $as, $($x)*);
    };
    ($l:lifetime $as:expr, unit $($x:tt)*) => {
        fn visit_unit<E: de::Error>(self) -> Result<Self::Value, E> { Ok($as) }
        visit_unconditionally!($l $as, $($x)*);
    };
    ($l:lifetime $as:expr, newtype_struct $($x:tt)*) => {
        fn visit_newtype_struct<D: Deserializer<$l>>(self, _: D) -> Result<Self::Value, D::Error> { Ok($as) }
        visit_unconditionally!($l $as, $($x)*);
    };
    ($l:lifetime $as:expr, seq $($x:tt)*) => {
        fn visit_seq<A: de::SeqAccess<$l>>(self, _: A) -> Result<Self::Value, A::Error> { Ok($as) }
        visit_unconditionally!($l $as, $($x)*);
    };
    ($l:lifetime $as:expr, map $($x:tt)*) => {
        fn visit_map<A: de::MapAccess<$l>>(self, _: A) -> Result<Self::Value, A::Error> { Ok($as) }
        visit_unconditionally!($l $as, $($x)*);
    };
    ($l:lifetime $as:expr, enum $($x:tt)*) => {
        fn visit_enum<A: de::EnumAccess<$l>>(self, _: A) -> Result<Self::Value, A::Error> { Ok($as) }
        visit_unconditionally!($l $as, $($x)*);
    };
}

#[derive(Debug, PartialEq, Eq, Clone)]
enum BrowserField {
    Empty,
    Main(PathBuf),
    Complex(BrowserSubstitutionMap),
}

impl BrowserField {
    fn to_map(self, main: &Path) -> BrowserSubstitutionMap {
        match self {
            BrowserField::Empty => Default::default(),
            BrowserField::Main(mut to) => {
                if !to.is_explicitly_relative() {
                    to.prepend_resolving(Path::new("."));
                }
                BrowserSubstitutionMap(map! {
                    PathBuf::from(".") => BrowserSubstitution::Replace(to.clone()),
                    main.to_owned() => BrowserSubstitution::Replace(to.clone()),
                })
            }
            BrowserField::Complex(map) => map,
        }
    }
}
impl Default for BrowserField {
    fn default() -> Self {
        BrowserField::Empty
    }
}
impl<'de> Deserialize<'de> for BrowserField {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        struct BrowserFieldVisitor;

        impl<'de> Visitor<'de> for BrowserFieldVisitor {
            type Value = BrowserField;

            fn expecting(&self, f: &mut fmt::Formatter) -> fmt::Result {
                write!(f, "anything at all")
            }
            fn visit_map<A: de::MapAccess<'de>>(self, access: A) -> Result<Self::Value, A::Error> {
                Ok(BrowserField::Complex(Deserialize::deserialize(
                    de::value::MapAccessDeserializer::new(access),
                )?))
            }
            fn visit_str<E: de::Error>(self, v: &str) -> Result<Self::Value, E> {
                Ok(BrowserField::Main(PathBuf::from(v)))
            }

            visit_unconditionally!('de BrowserField::Empty, bool i64 i128 u64 u128 f64 bytes none some unit newtype_struct seq enum);
        }

        deserializer.deserialize_any(BrowserFieldVisitor)
    }
}

// The main might be a string or an array of strings
fn from_main<'de, T, D>(deserializer: D) -> Result<Option<T>, D::Error>
where
    for<'a> T: From<&'a str> + Deserialize<'de>,
    D: Deserializer<'de>,
{
    struct FromMain<T>(PhantomData<T>);

    impl<'de, T> Visitor<'de> for FromMain<T>
    where
        for<'a> T: From<&'a str> + Deserialize<'de>,
    {
        type Value = Option<T>;

        fn expecting(&self, f: &mut fmt::Formatter) -> fmt::Result {
            write!(f, "string or array or nothing")
        }

        fn visit_str<E: de::Error>(self, v: &str) -> Result<Self::Value, E> {
            Ok(Some(T::from(v)))
        }

        fn visit_seq<S>(self, mut seq: S) -> Result<Self::Value, S::Error>
        where
            S: SeqAccess<'de>,
        {
            let mut value: Option<T> = None;
            while let Ok(item) = seq.next_element::<String>() {
                if let Some(item) = item {
                    if item.ends_with(".js") {
                        value = Some(T::from(&item));
                    }
                } else {
                    break;
                }
            }
            Ok(value)
        }

        visit_unconditionally!('de None, bool i64 i128 u64 u128 f64 bytes none some unit newtype_struct map enum);
    }

    deserializer.deserialize_any(FromMain(PhantomData))
}

impl<'de, T> Deserialize<'de> for BrowserSubstitution<T>
where
    for<'a> T: From<&'a str>,
{
    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
        struct SubstitutionVisitor<T>(PhantomData<T>);

        impl<'de, T> Visitor<'de> for SubstitutionVisitor<T>
        where
            for<'a> T: From<&'a str>,
        {
            type Value = BrowserSubstitution<T>;

            fn expecting(&self, f: &mut fmt::Formatter) -> fmt::Result {
                write!(f, "substitution path or false")
            }

            fn visit_bool<E: de::Error>(self, v: bool) -> Result<Self::Value, E> {
                if v {
                    Err(de::Error::invalid_value(de::Unexpected::Bool(v), &self))
                } else {
                    Ok(BrowserSubstitution::Ignore)
                }
            }

            fn visit_str<E: de::Error>(self, v: &str) -> Result<Self::Value, E> {
                Ok(BrowserSubstitution::Replace(From::from(v)))
            }
        }

        deserializer.deserialize_any(SubstitutionVisitor(PhantomData))
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use matches::assert_matches;

    #[test]
    fn test_deserialize_browser_subst() {
        let parse = serde_json::from_str::<BrowserSubstitution<String>>;
        assert_matches!(parse("null"), Err(_));
        assert_matches!(parse("100"), Err(_));
        assert_matches!(parse("[1, 2, 3]"), Err(_));
        assert_matches!(parse("false"), Ok(BrowserSubstitution::Ignore));
        assert_matches!(parse("true"), Err(_));
        assert_eq!(
            parse(r#""asdf""#).unwrap(),
            BrowserSubstitution::Replace("asdf".to_owned())
        );
        assert_eq!(
            parse(r#""""#).unwrap(),
            BrowserSubstitution::Replace("".to_owned())
        );
    }

    #[test]
    fn test_deserialize_browser() {
        let parse = serde_json::from_str::<BrowserSubstitutionMap>;
        assert_matches!(parse(r#"null"#), Err(_));
        assert_matches!(parse(r#""simple.browser.js""#), Err(_));
        assert_eq!(parse(r#"{}"#).unwrap(), BrowserSubstitutionMap(map! {}));
        assert_eq!(
            parse(r#"{"mod": "dom"}"#).unwrap(),
            BrowserSubstitutionMap(map! {
                PathBuf::from("mod") => BrowserSubstitution::Replace(PathBuf::from("dom")),
            })
        );
        assert_eq!(
            parse(r#"{"./file.js": "./file.browser.js"}"#).unwrap(),
            BrowserSubstitutionMap(map! {
                PathBuf::from("./file.js") => BrowserSubstitution::Replace(PathBuf::from("./file.browser.js")),
            })
        );
        assert_eq!(
            parse(r#"{"ignore": false}"#).unwrap(),
            BrowserSubstitutionMap(map! {
                PathBuf::from("ignore") => BrowserSubstitution::Ignore,
            })
        );
        assert_eq!(
            parse(
                r#"{
        "ignore": false,
        "mod": "dom",
        "mod2file": "./modfile.js",
        "mod2up": "../up.js",
        "mod2dir": "./moddir",
        "mod2abs": "/z/y/x",
        "./fileignore.js": false,
        "./file2mod.js": "mod",
        "./file2file.js": "./file.js",
        "./file2dir.js": "./dir",
        "./file2up.js": "../up.js",
        "./file2abs.js": "/x/y/z"
    }"#
            )
            .unwrap(),
            BrowserSubstitutionMap(map! {
                PathBuf::from("ignore") => BrowserSubstitution::Ignore,
                PathBuf::from("mod") => BrowserSubstitution::Replace(PathBuf::from("dom")),
                PathBuf::from("mod2file") => BrowserSubstitution::Replace(PathBuf::from("./modfile.js")),
                PathBuf::from("mod2up") => BrowserSubstitution::Replace(PathBuf::from("../up.js")),
                PathBuf::from("mod2dir") => BrowserSubstitution::Replace(PathBuf::from("./moddir")),
                PathBuf::from("mod2abs") => BrowserSubstitution::Replace(PathBuf::from("/z/y/x")),
                PathBuf::from("./fileignore.js") => BrowserSubstitution::Ignore,
                PathBuf::from("./file2mod.js") => BrowserSubstitution::Replace(PathBuf::from("mod")),
                PathBuf::from("./file2file.js") => BrowserSubstitution::Replace(PathBuf::from("./file.js")),
                PathBuf::from("./file2dir.js") => BrowserSubstitution::Replace(PathBuf::from("./dir")),
                PathBuf::from("./file2up.js") => BrowserSubstitution::Replace(PathBuf::from("../up.js")),
                PathBuf::from("./file2abs.js") => BrowserSubstitution::Replace(PathBuf::from("/x/y/z")),
            })
        );
    }

    #[test]
    fn test_deserialize_package_info() {
        let parse = serde_json::from_str::<PackageInfo>;
        assert_matches!(parse("null"), Err(_));
        assert_matches!(parse("100"), Err(_));
        assert_matches!(parse("[1, 2, 3]"), Err(_));
        assert_eq!(
            parse(r#"{}"#).unwrap(),
            PackageInfo {
                main: PathBuf::from("./index"),
                browser_substitutions: BrowserSubstitutionMap(map! {}),
            }
        );
        assert_eq!(
            parse(r#"{"browser": null}"#).unwrap(),
            PackageInfo {
                main: PathBuf::from("./index"),
                browser_substitutions: BrowserSubstitutionMap(map! {}),
            }
        );
        assert_eq!(
            parse(r#"{"browser": "simple"}"#).unwrap(),
            PackageInfo {
                main: PathBuf::from("./index"),
                browser_substitutions: BrowserSubstitutionMap(map! {
                    PathBuf::from(".") => BrowserSubstitution::Replace(PathBuf::from("./simple")),
                }),
            }
        );
        assert_eq!(
            parse(r#"{"browser": {}}"#).unwrap(),
            PackageInfo {
                main: PathBuf::from("./index"),
                browser_substitutions: BrowserSubstitutionMap(map! {}),
            }
        );
        assert_eq!(
            parse(r#"{"browser": {"mod": false}}"#).unwrap(),
            PackageInfo {
                main: PathBuf::from("./index"),
                browser_substitutions: BrowserSubstitutionMap(map! {
                    PathBuf::from("mod") => BrowserSubstitution::Ignore,
                }),
            }
        );
    }
}
