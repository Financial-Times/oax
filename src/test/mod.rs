// TODO move unit tests to relevant files (or folders)
#![allow(unused_imports)]

#[cfg(feature = "bench")]
extern crate test;

use super::*;
use fnv::{FnvHashMap, FnvHashSet};
use indoc::indoc;
use input_options::*;
use insta::assert_snapshot;
use manifest::{BrowserSubstitution, BrowserSubstitutionMap, PackageInfo};
use matches::assert_matches;
use path_ext::*;
use serde_json;
use std::io::{self, Write};
use std::path::Path;
use std::{ffi, fs, process};
use vlq::Vlq;
use walkdir::WalkDir;

#[test]
fn test_bundle_snapshots() {
    let entry_point = Path::new("examples/one-file/index.js");
    let options = InputOptions::default();
    let output = "examples/one-file/bumble.js";
    let map_output = SourceMapOutput::Suppressed;
    let _ = bundle(&entry_point, options, &output, &map_output).unwrap();
    assert_snapshot!(
        "one-file bundle",
        std::fs::read_to_string("examples/one-file/bumble.js").unwrap()
    );
}

#[test]
fn test_count_lines() {
    assert_eq!(count_lines(""), 1);
    assert_eq!(count_lines("this is a line"), 1);
    assert_eq!(count_lines("this is a line\n"), 2);
    assert_eq!(count_lines("\nthis is a line"), 2);
    assert_eq!(count_lines("\n\n\nthis is a line"), 4);
    assert_eq!(count_lines("this is a line\n\n\n"), 4);
    assert_eq!(count_lines("these\nare\nlines"), 3);
    assert_eq!(count_lines("\r\n"), 2);
    assert_eq!(count_lines("this is a line\r\n"), 2);
    assert_eq!(count_lines("\r\nthis is a line"), 2);
    assert_eq!(count_lines("these\nare\r\nlines"), 3);
}

#[test]
fn test_gather_npm_dev_deps_finds_complicated_deps() {
    let mut expected = FnvHashSet::<String>::default();
    let names = [
        "@with-org-name/nested",
        "optional",
        "non-optional",
        "yay",
        "mutual-a",
        "self",
        "circular",
        "@with-org-name/root",
        "mutual-b",
    ];
    for package in names.iter() {
        expected.insert(package.to_string());
    }
    let found =
        gather_npm_dev_deps(&"examples/npm-dev-dep-with-complicated-deps/index.js".to_owned())
            .unwrap();
    let found_deep_entry =
        gather_npm_dev_deps(&"examples/npm-dev-dep-with-complicated-deps/deep/entry.js".to_owned())
            .unwrap();
    assert_eq!(expected, found);
    assert_eq!(
        expected, found_deep_entry,
        "didn't go up looking for the nearest package.json"
    );
}

#[test]
fn test_gather_npm_dev_deps_doesnt_fail_on_missing_optionals() {
    assert!(
        gather_npm_dev_deps(
            &"examples/npm-dev-dep-with-missing-optional-deps/index.js".to_owned(),
        )
        .is_ok(),
        "failed, tried to analyse a missing optional dependency. should have ignored it",
    );
}

#[test]
fn test_gather_npm_dev_deps_fails_on_missing_required_deps() {
    assert!(
        gather_npm_dev_deps(
            &"examples/npm-dev-dep-with-missing-required-deps/index.js".to_owned(),
        )
        .is_err(),
        "skipped a required (non-optionalDependencies) dependency",
    );
}

use cfg_if::cfg_if;
cfg_if! {
    if #[cfg(feature = "bench")] {
        #[bench]
        fn bench_vlq(b: &mut test::Bencher) {
            let mut vlq = Vlq::new();
            b.iter(|| {
                test::black_box(vlq.enc(-1001));
            });
        }

        #[bench]
        fn bench_cjs_simple(b: &mut test::Bencher) {
            let entry_point = Path::new("examples/simple/index.js");
            npm_install(entry_point.parent().unwrap());
            let options = InputOptions::default();
            let output = "/dev/null";
            let map_output = SourceMapOutput::Inline;

            b.iter(|| {
                let _ = bundle(&entry_point, options, &output, &map_output).unwrap();
            });
        }

        #[bench]
        fn bench_es6_simple(b: &mut test::Bencher) {
            let entry_point = Path::new("examples/es6-simple/index.mjs");
            npm_install(entry_point.parent().unwrap());
            let options = InputOptions::default();
            let output = "/dev/null";
            let map_output = SourceMapOutput::Inline;

            b.iter(|| {
                let _ = bundle(&entry_point, options, &output, &map_output).unwrap();
            });
        }

        #[bench]
        fn bench_es6_everywhere_simple(b: &mut test::Bencher) {
            let entry_point = Path::new("examples/es6-everywhere-simple/index.js");
            npm_install(entry_point.parent().unwrap());
            let options = InputOptions::default();
            let output = "/dev/null";
            let map_output = SourceMapOutput::Inline;

            b.iter(|| {
                let _ = bundle(&entry_point, options, &output, &map_output).unwrap();
            });
        }

        #[bench]
        fn bench_write_map_to(b: &mut test::Bencher) {
            let writer = Writer {
                modules: {
                    let mut modules = FnvHashMap::default();
                    for i in 0..1000 {
                        let mut path = PathBuf::new();
                        path.push(i.to_string());
                        path.push("examples/es6-everywhere-simple/node_modules/itt/index.js");
                        modules.insert(
                            path,
                            Module {
                                source: Source {
                                    prefix: "~function() {".to_owned(),
                                    body: include_str!("itt.js").to_owned(),
                                    suffix: "}()".to_owned(),
                                    original: None,
                                },
                                deps: {
                                    let mut deps = FnvHashMap::new();
                                    deps.insert("./math".to_owned(), Resolved::Normal(
                                        Path::new("examples/es6-everywhere-simple/math.js").to_owned(),
                                    ));
                                    deps.insert("itt".to_owned(), Resolved::Normal(
                                        Path::new("examples/es6-everywhere-simple/node_modules/itt/index.js").to_owned(),
                                    ));
                                    deps
                                },
                            },
                        );
                    }
                    modules
                },
                entry_point: Path::new("examples/es6-everywhere-simple/index.js"),
                map_output: &SourceMapOutput::Inline,
            };

            let mut out = Vec::new();
            b.iter(|| {
                out.clear();
                writer.write_map_to(&mut out).unwrap();
            });
            b.bytes = out.len() as u64;
        }
    }
}
