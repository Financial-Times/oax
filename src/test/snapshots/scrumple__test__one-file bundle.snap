---
source: src/test/mod.rs
expression: "std::fs::read_to_string(\"examples/one-file/bumble.js\").unwrap().replace(\"\\r\\n\",\n                                                                        \"\\n\")"
---
(function() {
  var Scrumple = {}
  Scrumple.baseRequire = typeof require !== "undefined" ? require : function(n) {
    throw new Error("Could not resolve module name: " + n)
  }
  Scrumple.ignored = function(){}
  Scrumple.ignored.deps = {}
  Scrumple.ignored.filename = ''
  Scrumple.modules = {}
  Scrumple.files = {}
  Scrumple.mains = {}
  Scrumple.resolve = function (base, then) {
    base = base.split('/')
    base.shift()
    then.split('/').forEach(function(p) {
      if (p === '..') base.pop()
      else if (p !== '.') base.push(p)
    })
    return '/' + base.join('/')
  }
  Scrumple.Module = function Module(filename, parent) {
    this.filename = filename
    this.id = filename
    this.loaded = false
    this.parent = parent
    this.children = []
    this.exports = {}
  }
  Scrumple.makeRequire = function (self) {
    var require = function(m) { return require._module(m).exports }
    require._deps = {}
    require.main = self

    require._esModule = function (m) {
      var mod = require._module(m)
      return mod.exports.__esModule ? mod.exports : {
        get default() {return mod.exports}
      }
    }
    require._module = function (m) {
      var fn = self ? require._deps[m] : Scrumple.main
      if (fn == null) {
        var module = {exports: Scrumple.baseRequire(m)}
        require._deps[m] = {module: module}
        return module
      }
      if (fn.module) return fn.module
      var module = new Scrumple.Module(fn.filename, self)
      fn.module = module
      module.require = Scrumple.makeRequire(module)
      module.require._deps = fn.deps
      module.require.main = self ? self.require.main : module
      if (self) self.children.push(module)
      fn(module, module.exports, module.require, fn.filename, fn.filename.split('/').slice(0, -1).join('/'), {url: 'file://' + (fn.filename.charAt(0) === '/' ? '' : '/') + fn.filename})
      module.loaded = true
      return module
    }
    return require
  }

  Scrumple.files["index.js"] = file_index$2ejs; file_index$2ejs.deps = {}; file_index$2ejs.filename = "index.js"; function file_index$2ejs(module, exports, require, __filename, __dirname, __import_meta) {
Object.defineProperty(exports, '__esModule', {value: true})
; ~function() {
'use strict';
Object.defineProperties(exports, {
  default: {get() {return __default}, enumerable: true, configurable: true},
  named: {get() {return named}, enumerable: true, configurable: true},
});

 const __default =  {}
 let named = "named"
}()}
  Scrumple.main = file_index$2ejs; Scrumple.makeRequire(null)()
  if (typeof module !== 'undefined') module.exports = Scrumple.main.module && Scrumple.main.module.exports
})()

